enhancements/cluster-logging/cluster-logging-v2-apis.md

```diff
@@ -0,0 +1,429 @@
+* Support log forwarder API with minimal dependency upon reserved words (e.g. default)
+* Support an API to spec a Red Hat managed LokiStack with the logging tenancy model
+* Support an API to allow flexible deployment of the logging components: collector/forwarder, storage, visualization
+* Continue to allow deployment of a log forwarder to the output sinks of the administrators choosing
+
+
+### Non-Goals
+
+* "One click" deployment of a full logging stack as provided by **ClusterLogging** v1
+* Complete backwards compatibility to **ClusterLogForwarder** v1
+* Automated migration path from v1 to v2
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal
+actually is. Describe clearly what will be changed, including all of
+the components that need to be modified and how they will be
+different. Include the reason for each choice in the design and
+implementation that is proposed here, and expand on reasons for not
+choosing alternatives in the Alternatives section at the end of the
+document.
+
+### Workflow Description
+
+The following workflow describes the first user story which is a superset of the others and allows deployment of a full logging stack to collect and forward logs to a Red Hat managed log store.
+
+**cluster administrator** is a human responsible for:
+
+* Managing and deploying day 2 operators
+* Managing and deploying an on-cluster LokiStack
+* Managing and deploying a cluster-wide log forwarder
+
+**obervability-operator** is an operator responsible for:
+
+* managing and deploying observability plugins (e.g log-view-plugin)
+
+**loki-operator** is an operator responsible for managing a loki stack
+
+**cluster-logging-operator** is an operator responsible for managing log collection and forwarding
+
+The cluster administrator does the following:
+
+1. Deploys the Red Hat **observability-operator**
+1. Deploys the Red Hat **loki-operator**
+1. Deploys an instance of **LokiStack** in the `openshift-logging` namespace
+1. Deploys the Red Hat **cluster-logging-operator**
+1. Creates a **ClusterLogForwarder** custom resource for the **LokiStack**
+
+The **loki-operator**:
+
+1. Deploys the logging-view-plugin for reading logs in the OpenShift console
+
+The **cluster-logging-operator**:
+
+1. Deploys the log collector to forward logs to log storage in the `openshift-logging` namespace
+
+### API Extensions
+
+CustomResourceDefinition:
+```yaml
+    apiVersion: "logging.openshift.io/v2"
+    kind: ClusterLogForwarder
+    metadata:
+      name: 
+    spec:
+      serviceAccount:
+        name: 
+        namespace:       #namespace of deployment and resources
+      inputs:
+      - name:
+        type:
+        application:
+          selector:  #labelselector
+          namespaces:
+            include: []
+            exclude: []
+          containers:
+            include: []
+            exclude: []
+            maxRecordsPerSecond: # map[containername] int 
```

Comment:
- For future-proofing, suggest a "limits" section, same as outputs, even if there is only one limit now. - make the name more explict, example below isn't ideal we can brainstorm
- need a wildcard for containers that aren't named explicitly

```
  limits:
    maxRecordsPerSecondPerContainer:  # too long, brainstorm
       ngnix: 100
       any : 50
```

Above: ngnix containers get 100 r/s, any other container else gets 50.
"*" is a natural wildcard but using in JSON keys is asking for trouble
Don't think we can/should support pattern matching (wildcard or regex), only exact match or "anything goes"

Reply:
We already have an issue with this value an only be applied per node deployment but the CLF reads like it applies as the aggregate.  I would say there is no good name that is not long or difficult.  I think we can only resolve this by API documentation:
```
containers:
  limits:
   maxRecords:    #or maxRecordsSec or MaxRecordsPerSec
       mysidecare: 100
```

Does :arrow_up:  align with your comment or are you suggesting a limits section that is a child to outputs?

> Above: ngnix containers get 100 r/s, any other container else gets 50.

I would disagree with magic 'any'.  Admin can spec an alternate inputs to capture "any"

> "*" is a natural wildcard but using in JSON keys is asking for trouble

I agree we do not support wild card given vector only allows an explicit key.  The wild card will introduce complications to the implementation.

Reply:
Based on discussoin, add a tuning section for consistency with output and for future proof:
```
inputs:
  - tuning:      
        rateLimitDefault:
	        recordsPerSecond: #int - for containers not mentioned in rateLimitByContainer
	        # policy: Drop|Block # may add in future, currently we only drop.
        rateLimitByContainer:  # map[string]RateLimit - limits by container name.
          # Example
          ngnix: 
  	        recordsPerSecond: 50
    	      # policy: Drop|Block # may add in future, currently we only drop.

Reply:
@alanconway please confirm the change I added